doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+Panda3D C++ Style Guide
+=======================
+
+Rather than try to justify every decision made, this document attempts to be as
+at-a-glance as possible.  Examples provided where necessary to illustrate best
+practices.
+
+Note that this style guide is, for the most part, not very strict, and there
+are exceptions to every rule.  However, this guide does serve to illustrate the
+preferred style for the entire codebase.
+
+Naming conventions
+------------------
+
+- **Classes**: `CamelCaseFirstUpper`
+- **Filenames**: `lowerCamelCase.h`, `lowerCamelCase.cxx`
+  - Note, every header file must have a unique filename.  They are all
+    installed into the same path.
+- **Functions**: `lower_snake_case_without_prefix`
+- **Local variables**: `lower_snake_case_without_prefix`
+- **Member variables**: `_lower_snake_case_with_initial_score`
+
+Indentation
+-----------
+
+Always use two spaces to indent.  Tabs should never appear in a C++ file.
+
+Access modifiers (`public`/`private`/`protected`/`PUBLISHED`) are never
+indented, and do not affect the indentation of the surrounding lines.
+
+Spaces
+------
+
+Always put one space after a control keyword, before the condition's
+parentheses.  Don't pad the inside of parentheses with spaces unless there are
+several nested parenthetical expressions and the extra spaces contribute
+substantially to readability.
+
+    if (x == y) {
+
+Don't place a space after a function name (declarations, definitions, or calls).
+
+    void function_name();
+    ...
+    function_name();
+
+Always place a space after each comma in an arguments list.
+
+    function_call(x, y, z);
+
+Always pad binary operators with a space on either side.
+
+    a || b
+
+Never end a line with trailing whitespace.
+
```

Comment:
I would propose adding something explicitly forbidding [assignment alignment](https://www.reddit.com/r/readablecode/comments/19vkkz/do_you_guys_align_your_code_into_columns/).  While arguably being readable to align a bunch of assignments, it is a pain in the long run to constantly have to realign blocks of variables or function calls whenever something is changed.

Reply:
That'd be a good thing to put under the "Alignment" section.

For what it's worth, I have a WIP `.clang-format` file (and a script to temporarily edit some of the Panda macros into more standard-looking C++ so clang-format can understand it) located [here](https://gist.github.com/CFSworks/42be608c5ead2251542163e3b97b66b2). It does do a very good job of aligning everything for you - and it has several editor integrations so we can tell people to use that (and perhaps even have Travis fail the PR if it doesn't pass clang-tidy/clang-format checks).

If we want to explicitly _forbid_ it, that's fine too - clang-format also does a great job of unifying everything to a single space.

The thing it doesn't seem to support is our odd habit of always breaking function _definitions_ to align the function name to column 1.

Reply:
I think I would agree to rdb, while aligment can be useful in some cases, in many cases it is just annoying to maintain.